File: leetcode/stacks/lc_20_valid_parentheses.py
'''
https://leetcode.com/problems/valid-parentheses/description/
'''
class Solution:
    def isValid(self, s: str) -> bool:
        # create a dict that maps
        # close bracket to open bracket
        close_to_open = {
            ')': '(',
            ']': '[',
            '}': '{'
        }

        # create our bracket stack
        # this stack will be used to
        # add open brackets to
        bracket_stack = []

        # loop through all characters in s
        for char in s:
            # if the character is an open bracket
            if char not in close_to_open:
                # then append to the bracket stack
                bracket_stack.append(char)
            # else the character is a close bracket
            else:
                # if the bracket stack is empty
                # meaning there is no open bracket
                # left to close, then we have an error
                if not bracket_stack:
                    # so return False
                    return False

                # so get the top bracket from the stack
                top_bracket = bracket_stack.pop()

                # if the top bracket does not match this
                # close bracket, then we have an error
                if top_bracket != close_to_open[char]:
                    # so return False
                    return False
        
        # at the very end, we want the bracket stack
        # to be empty, so only return True if it's empty
        return not bracket_stack

File: leetcode/cpp/stack/lc_20_valid_parentheses.cpp
#include <stack>
#include <string>

/**
 * https://leetcode.com/problems/valid-parentheses/description/
 *
 * Time complexity: O(n) where n = length of s
 * Space complexity: O(n) where n = length of s
 */
class Solution {
public:
    bool isValid(std::string s) {
        // if there are less than 2 chars in s
        // automatically valid parentheses cannot be possible
        // so return false
        if (s.size() < 2) {
            return false;
        }

        // setup our stack
        std::stack<char> stringStack;

        // loop through all characters in s
        for (const char& chr: s) {
            // for every closing bracket, return false if any of the following:
            // 1) the correct closing bracket doesn't exist at the top of the stack
            // 2) the stack is empty
            // otherwise continue and add every opening bracket to the stack
            switch(chr) {
                case ')':
                    if (stringStack.empty() || stringStack.top() != '(') {
                        return false;
                    }
                    stringStack.pop();
                    break;
                case ']':
                    if (stringStack.empty() || stringStack.top() != '[') {
                        return false;
                    }
                    stringStack.pop();
                    break;
                case '}':
                    if (stringStack.empty() || stringStack.top() != '{') {
                        return false;
                    }
                    stringStack.pop();
                    break;
                default:
                    stringStack.push(chr);
                    break;
            }
        }

        // at the end, make sure that the stack is empty
        // so all opening brackets have been closed
        // and then return true else false
        return stringStack.empty();
    }
};
